### PR TITLE
IntegrateEllipsoid now use separate integrators for Bragg peaks and satellite peaks

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

- Original Gitlab user story: [SCD369](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/369)
- This is #32008 into `master`

The original `IntegrateEllipsoid` uses a single integrator to process all peaks in given workspace, which forced the users to

- separate Bragg peaks and satellite peaks prior to integration
- run integration for the two set of peaks independently
- merge the results back into a single workspace

This PR introduces one additional integrator (for satellite peaks) such that

- if no satellite peak options are specified -> treat satellite peaks the same as Bragg peaks
- else -> use `integrator_satellite` to integrate satellite peaks while using `integrator` for regular Bragg peaks.

This new feature help simplify the workflow on the user end.

**To test:**

The testing data is available on SNS netdrive, therefore it is recommended to perform the test there

```python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
​
Load(Filename='/SNS/TOPAZ/IPTS-23996/nexus/TOPAZ_36079.nxs.h5', 
     OutputWorkspace='TOPAZ_36079.nxs_event',
     FilterByTofMin=500, 
     FilterByTofMax=16666)
​
FilterBadPulses(InputWorkspace='TOPAZ_36079.nxs_event', 
                OutputWorkspace='TOPAZ_36079.nxs_event', 
                LowerCutoff=25)
​
LoadIsawDetCal(InputWorkspace='TOPAZ_36079.nxs_event', 
               Filename='/SNS/TOPAZ/IPTS-23996/shared/calibration/TOPAZ_2020A.DetCal')
​
CropToComponent('TOPAZ_36079.nxs_event', ComponentNames='bank13', OutputWorkspace='crop')
​
DeleteWorkspace("TOPAZ_36079.nxs_event")
​
detids = [858225, 905237, 874008]
​
ExtractSpectra(InputWorkspace="crop",
               OutputWorkspace="crop",
               DetectorList="858220-858230, 905232-905242, 874003-874013") # +/- 5 detectors around each detid

UB = np.array([[ 0.15468228,  0.10908475, -0.14428671],
               [-0.08922105, -0.08617147, -0.22976459],
               [-0.05616441,  0.12536522, -0.03238277]])
​
ConvertToMD(InputWorkspace='crop', 
            QDimensions='Q3D', 
            dEAnalysisMode='Elastic',
            Q3DFrames='Q_sample', 
            LorentzCorrection=True, 
            OutputWorkspace='md', 
            MinValues='1,1,1.675', # closer crop to the data
            MaxValues='10,5,8.425')
            
CreatePeaksWorkspace(InstrumentWorkspace='crop', NumberOfPeaks=0, OutputWorkspace='peaks')
​
SetUB('peaks', UB=UB)
​
AddPeakHKL('peaks', [0.15, 1.85, -1])
AddPeakHKL('peaks', [1, 4, -3])
AddPeakHKL('peaks', [1, 5, -3])
​
IndexPeaks(PeaksWorkspace='peaks', 
           Tolerance=0.06, 
           ToleranceForSatellite=0.05, 
           RoundHKLs=False, 
           ModVector1='0.125,0,0', 
           ModVector2='0,0.125,0', 
           ModVector3='-0.125,0.125,0', 
           MaxOrder=1, 
           CrossTerms=False, 
           SaveModulationInfo=True)

# treat satellite peaks as if they were Bragg peaks
IntegrateEllipsoids(InputWorkspace='crop', 
                    PeaksWorkspace='peaks', 
                    RegionRadius=0.055, 
                    SpecifySize=True, 
                    PeakSize=0.0425,
                    BackgroundInnerSize=0.043, 
                    BackgroundOuterSize=0.055, 
                    OutputWorkspace='peaks_integrated_nosatellite', 
                    CutoffIsigI=5,
                    UseOnePercentBackgroundCorrection=False)

# provide satellite peak specific options
IntegrateEllipsoids(InputWorkspace='crop', 
                    PeaksWorkspace='peaks', 
                    RegionRadius=0.055, 
                    SpecifySize=True, 
                    PeakSize=0.0425,
                    BackgroundInnerSize=0.043, 
                    BackgroundOuterSize=0.055, 
                    OutputWorkspace='peaks_integrated_satellite', 
                    CutoffIsigI=5,
                    UseOnePercentBackgroundCorrection=False,
                    SatelliteRegionRadius=0.1, 
                    SatellitePeakSize=0.08, 
                    SatelliteBackgroundInnerSize=0.081, 
                    SatelliteBackgroundOuterSize=0.1)
​
​# check results
peaks_nosat = mtd["peaks_integrated_nosatellite"]
peaks_sat   = mtd["peaks_integrated_satellite"]
​
print("testing that properties were set...")
assert peaks_sat.run().hasProperty("PeaksIntegrated")
assert peaks_sat.run().hasProperty("SatellitePeakRadius")
assert peaks_sat.run().hasProperty("SatelliteBackgroundInnerRadius")
assert peaks_sat.run().hasProperty("SatelliteBackgroundOuterRadius")
​
n = peaks_nosat.getNumberPeaks()
for i in range(n):
    peak = peaks_nosat.getPeak(i)
    satpeak = peaks_sat.getPeak(i)
    
    peak_mnp = peak.getIntMNP()
    satpeak_mnp = satpeak.getIntMNP()
    
    peak_int = peak.getIntensity()
    satpeak_int = satpeak.getIntensity()
    
    print("peak {}:".format(i+1))
    print("  NO satellite:   MNP = {}, intensity = {}".format(peak_mnp, peak_int))
    print("  WITH satellite: MNP = {}, intensity = {}".format(satpeak_mnp, satpeak_int))
    
    # Verify that intensities match for non-satellite peaks
    if satpeak_mnp.norm2() == 0.0:
        np.testing.assert_almost_equal(peak_int, satpeak_int)
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by jmborr** [comment]
@KedoKudo I think the documentation of `IntegrateEllipsoid` should contain a paragraph explaining the case when you want to differentiate between main and satellite peaks, and the case when you don't want that (for instance, you may have no _a priori_ knowledge of the modulation vectors). You could use parts of the workbench test script to showcase the two "execution modes". There's no need for the inserted code to become a doctests, thought, just to show how the two modes are called by the algorithm

**Comment by KedoKudo** [comment]
> @KedoKudo I think the documentation of `IntegrateEllipsoid` should contain a paragraph explaining the case when you want to differentiate between main and satellite peaks, and the case when you don't want that (for instance, you may have no _a priori_ knowledge of the modulation vectors). You could use parts of the workbench test script to showcase the two "execution modes". There's no need for the inserted code to become a doctests, thought, just to show how the two modes are called by the algorithm

somehow my `git rebase` excluded the documentation updates :sweat: 

Now the PR should contain the updated documentation with examples.

**Comment by jmborr** [comment]
I ran your workbench script and I'm a bit confused. The reported intensities for the satellite peak are as follows:
```bash
peak 1:
  NO satellite:   MNP = [0,0,-1], intensity = 17.00000000100127
  WITH satellite: MNP = [0,0,-1], intensity = 13.000000000500636
```
So the satellite peak is assigned more intensity when interpreted as a main peak. However, if I look at the various geometrical boundaries in the two scenarios:
```python
# Treat the satellite peak as a main peak
IntegrateEllipsoids(...RegionRadius=0.055, PeakSize=0.0425,
                    BackgroundInnerSize=0.043,  BackgroundOuterSize=0.055...)
# Tread the peak as satellite
IntegrateEllipsoids(...SatelliteRegionRadius=0.1,  SatellitePeakSize=0.08,
                    SatelliteBackgroundInnerSize=0.081, SatelliteBackgroundOuterSize=0.1...)
```
looks like the scenario that considers the satellite as a main Bragg peak  assigns a smaller Q-volume (0.0425) than the scenario that explicitly considers the peak as a satellite (0.08). Thus, I was expecting a smaller intensity in the first scenario since there should be less events to be collected because of the smaller Q-volume. Could the disparity be caused by the subtraction of the different backgrounds?

**Comment by KedoKudo** [comment]
> I ran your workbench script and I'm a bit confused. The reported intensities for the satellite peak are as follows:
> 
> ```shell
> peak 1:
>   NO satellite:   MNP = [0,0,-1], intensity = 17.00000000100127
>   WITH satellite: MNP = [0,0,-1], intensity = 13.000000000500636
> ```
> 
> So the satellite peak is assigned more intensity when interpreted as a main peak. However, if I look at the various geometrical boundaries in the two scenarios:
> 
> ```python
> # Treat the satellite peak as a main peak
> IntegrateEllipsoids(...RegionRadius=0.055, PeakSize=0.0425,
>                     BackgroundInnerSize=0.043,  BackgroundOuterSize=0.055...)
> # Tread the peak as satellite
> IntegrateEllipsoids(...SatelliteRegionRadius=0.1,  SatellitePeakSize=0.08,
>                     SatelliteBackgroundInnerSize=0.081, SatelliteBackgroundOuterSize=0.1...)
> ```
> 
> looks like the scenario that considers the satellite as a main Bragg peak assigns a smaller Q-volume (0.0425) than the scenario that explicitly considers the peak as a satellite (0.08). Thus, I was expecting a smaller intensity in the first scenario since there should be less events to be collected because of the smaller Q-volume. Could the disparity be caused by the subtraction of the different backgrounds?

Yes, the instrument scientists would like the satellite peaks and Bragg peaks use completely different integration parameters, which includes how background is estimated.

**Review by jmborr — CHANGES_REQUESTED** [review]
Very very minor requests

File: docs/source/release/v6.2.0/diffraction.rst [hunk 1]
```diff
@@ -14,6 +14,7 @@ Powder Diffraction
 New features
 ############
 - New algorithm :ref:`CombineDiffCal <algm-CombineDiffCal>` to calibrate groups of pixels after cross correlation so that diffraction peaks can be adjusted to the correct positions
+- New script for doing calibration by groups, :ref:`PowderDiffractionCalibration <calibration_tofpd_group_calibration-ref>`
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply docs/source/release/v6.2.0/diffraction.rst:17]: It is under the improvements in the release note

File: Framework/MDAlgorithms/test/IntegrateEllipsoidsTest.h [hunk 2]
```diff
@@ -399,6 +401,143 @@ class IntegrateEllipsoidsTest : public CxxTest::TestSuite {
...
   }
+
+  void test_execution_background_shell() {
+
+    const std::vector<double> ub = {0.15468228,  0.10908475,  -0.14428671, -0.08922105, -0.08617147,
+                                    -0.22976459, -0.05616441, 0.12536522,  -0.03238277};
+
+    auto loadalg = AlgorithmManager::Instance().createUnmanaged("LoadNexusProcessed");
+    TS_ASSERT_THROWS_NOTHING(loadalg->initialize());
+    TS_ASSERT_THROWS_NOTHING(loadalg->setPropertyValue("Filename", "TOPAZ_36079_crop.nxs"));
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply Framework/MDAlgorithms/test/IntegrateEllipsoidsTest.h:412]: the testing data has already been cropped (thanks to @ckendrick), so the overall testing time is very short.

File: docs/source/algorithms/IntegrateEllipsoids-v1.rst [hunk 3]
```diff
@@ -164,6 +167,13 @@ ellipsoid. The outer surface of the background ellipsoidal shell is an
 ellipsoidal surface with the same relative axis lengths as the inner
 surface.
 
+By default, the algorithm does not distinguish satellite peaks from Bragg
+peaks by using identical integrators for two different type of peaks.
+However, users can specify *SatelliteRegionRadius*, *SatellitePeakSize*,
+*SatelliteBackgroundInnerSize* and *SatelliteBackgroundOuterSize* such that
+the integrator used for staellite peaks are different from the one used for
```
Comment (jmborr) [inline docs/source/algorithms/IntegrateEllipsoids-v1.rst:174]: typo in "staellite"

File: docs/source/algorithms/IntegrateEllipsoids-v1.rst [hunk 4]
```diff
@@ -285,6 +290,52 @@ file is available in `Mantid system tests repository <https://github.com/mantidp
    | 3132     | 1189484  |     -4.0 |     -1.0 |      6.0 |      1.1 |     63.4 |   5299.3 |      1.0 |  13512.0 |    120.7 |     31.0 | bank18   |    108.0 |     38.0 | [4.02414,3.39659,3.83664] | [6.4679,0.298896,0.726133] | 9        |
    | 3132     | 1218337  |     -5.0 |     -2.0 |      7.0 |      1.0 |     79.8 |   4724.1 |      0.8 |   7411.0 |     88.3 |     15.0 | bank18   |     33.0 |    151.0 | [4.96622,3.61607,5.32554] | [7.99244,1.19363,0.892655] | 10       |
 
+
+**Example - IntegrateEllipsoids with staellite peaks:**
```
Comment (jmborr) [inline docs/source/algorithms/IntegrateEllipsoids-v1.rst:294]: typo in "staellite"

**Review by jmborr — CHANGES_REQUESTED** [review]
Just two minuscule typos. Slack me after you submit the corrections

**Review by jmborr — APPROVED** [review]

**Pull request merged**